### DISCUSSION TITLE
Change default device and default communication device simultaneously

### DISCUSSION THREAD

**Discussion by haaskicker** [File-New-Project/EarTrumpet#682] [Ideas]
This issue is sort of a continuation of issue #71.

My problem mostly pertains to using Discord, but it does happen in other applications as well.
I like to use/talk on Discord using both my headphones (USB/DAC) and speakers (analog input) depending on if I need good audio for games, my ears need a rest, etc.

EarTrumpet is typically great for moving between the two, however, since my headphones are the "default communications device" on the computer, Discord will always play through them regardless of what the "default device" is (which is what EarTrumpet changes). 

If you guys could implement a settings option to change either/or/both the "default communications device" and "default device" then I think this problem would be solved.

**Comment by Luckz** [comment]
https://github.com/File-New-Project/EarTrumpet/issues/71#issuecomment-396823592 makes it sound like this was _resolved_ in version 2.0, but it isn't so?

EDIT: Updated Windows, verified it still doesn't do that.

**Comment by haaskicker** [comment]
> [#71 (comment)](https://github.com/File-New-Project/EarTrumpet/issues/71#issuecomment-396823592) makes it sound like this was _resolved_ in version 2.0, but it isn't so?

From what I've seen, there still isn't a solution. There is no option to manipulate the "default communication device" in addition to the "default device"

**Comment by davuxcom** [comment]
The comment in #71 refers to changing for Console AND Multimedia rendering, which isn't related to also changing the communication device at the same time.  We are aware of the request and this thread is tracking moving the comm device when moving the default device.

**Comment by VapidLinus** [comment]
**Example use case:** This is annoying in many games that use VOIP. A lot of them use the communication device for voice audio, while using the audio device for game audio. You generally want all sound to come from the audio device you're currently using - so whenever I want to switch between my audio devices, I still have to open the legacy sound menu to change both.

**Comment by Luckz** [comment]
> I still have to open the legacy sound menu to change both.

FWIW if you don't need the many other special features of EarTrumpet on a daily basis, both SoundSwitch and Audio Switcher can change devices for you much faster than the Sound Control Panel route. The latter way drove me crazy over time.
(Both can also be installed via scoop, the first tool from a third party bucket.)

**Comment by VapidLinus** [comment]
Thank you @Luckz. SoundSwitch was perfect for me. I will keep both installed.

**Comment by Aasikki** [comment]
This would be a nice thing to have. I play quite a bit in vr and having to go to discord or windows audio settings settings every time to change the output device is annoying. SoundSwitch works well but I still need EarTrumpet for the volume mixer, so having this option would eliminate having to use two applications.

**Comment by TIPConsulting** [comment]
Bump.  I would also really like this to set both (or all) default device outputs. Console, communication, and possibly media

**Comment by TIPConsulting** [comment]
Here is an implementation of a registry policy object that can change all 3 different audio device roles on windows 

https://github.com/frgnca/AudioDeviceCmdlets/blob/master/SOURCE/AudioDeviceCmdlets.cs#L412

I just integrated this code into a project, so I can confirm that this method works. It's fairly simple to extract the core functionality out of his commandlet implementation

**Comment by LukasDoubleU** [comment]
I implemented this and other features over at my fork.
There is a new setting to "change the default communication" device.
If enabled, EarTrumpet will change the comms device as well when changing the default output device.

If you want you can test the pre-release:
https://github.com/LukasDoubleU/EarTrumpet/releases

**Reply by pohuing** [reply]
Is there a reason this isn't in main yet? I've been looking to implement the same feature but if it's already done I'm not sure if I should bother.

**Reply by LukasDoubleU** [reply]
I have talked to the maintainer and my branch will not be merged into main.
Reason being, that there have been some changes in the background of the main that make my changes obsolete.
Not sure when those changes will be merged into main tho.

**Comment by pohuing** [comment]
It's been a year and the functionality hasn't been added yet. Did the changes get scrapped and @LukasDoubleU 's changes are relevant again?

**Reply by LukasDoubleU** [reply]
The change that made my changes obsolete as per the maintainer was the "actions" feature.
The idea was, that my enhancement proposals (change comm device, hide apps, ..) should be implemented as actions and therefore they were not merged as "native" features.
I myself will not be doing that anytime soon though as I do not have a C# dev environment.

**Reply by pohuing** [reply]
Well that's frustrating. I can't find any mention of such features in the docs nor in the settings. Guess I'll have to stick to applying your patch and just building myself occasionally. Thx

**Reply by e-t-l** [reply]
@pohuing could you describe briefly how you do that? (building Eartrumpet with @LukasDoubleU's patch?) I'm having the same issue and it seems like you've found a (reasonable) solution

**Reply by pohuing** [reply]
Back then I simply copied his changes as a patch from github and applied them in rider on my copy of the official repository. I'm not sure how much has changed upstream since then so idk how well that still works.

You can find the patch [here](https://github.com/File-New-Project/EarTrumpet/commit/2ea8718b93a3542d56acb5296099d953bd0331ed.diff). 
For learning how to apply a patch without rider(as that IDE costs money) you should consult the git docs or scour stackoverflow. 
You should ignore the changes in the app manifest and readme. The rest implements the new feature.

**Reply by e-t-l** [reply]
> back then

Do you not use eartrumpet anymore?

**Reply by pohuing** [reply]
I do, I just haven't had a need to update beyond my most recent build.

**Comment by DCaedus** [comment]
Yes please, I have a wireless Razer headset, wired earbuds, and my laptop's speakers, and I regularly use Discord to call with friends
Every time I want to use a different device, I have to go 

Right click EarTrumpet > Windows Legacy > Playback devices > Right click device once > Set as Default Device > Right click device again > Set as Default Communication device > Change to recording tab > Right click Mic once > Set as Default Device > Right click Mic again > Set as Default Communication device

After 15 clicks, I have achieved changing my audio device.
I know all this isn't caused by EarTrumpet, it's just how Windows works, but it would be insanely helpful to me and many other people, if mayb in EarTrumpet's settings there was a toggle/checkbox to; "When changing Default audio device, also change Default Communication device" perhaps.

**Comment by EnriadHodor** [comment]
This feature is still needed for many games and Discord. I don't want install a separate program like SoundSwitch just for one useful toggle. Please add it 🙏

**Comment by venelope-van-tabel** [comment]
Lord please

**Comment by Yootz** [comment]
bump, please add it
